File: app/src/main/java/com/dvor/my/mydvor/MessageFragment.java
package com.dvor.my.mydvor;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.Button;
import android.widget.TextView;

import java.util.*;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.Date;
import java.util.LinkedList;
import java.util.List;

public class MessageFragment extends Fragment implements View.OnClickListener{

    private static List<MyEventListener> eventListeners;

    public void addEventListener(MyEventListener eventListener) {
        eventListeners.add(eventListener);
    }

    public void notifyEventListeners(MyEvent event) {
        for (MyEventListener eventListener: eventListeners) {
            eventListener.processEvent(event);
        }
    }

    private FirebaseAuth mAuth;
    String userStreetId;
    String organizationId;
    String userBuildingId;
    DatabaseReference myRef2;
    DatabaseReference myRef3;
    DataSnapshot messageSnapshot;
    ValueEventListener listenerBuilding;
    ValueEventListener listenerMessages;

    private List<Message> messages = new ArrayList<Message>();
    private TextView messageText;
    ListView messageList;
    AdapterView.OnItemClickListener itemListener;
    Context context;

    //сохранение текста неотправленного сообщения
    //костыльно но работает
    public void onSaveInstanceState(Bundle outState) {
        outState.putString("message", (messageText.getText().toString()));
        MainActivity.savedMessage=messageText.getText().toString();
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);

        if(eventListeners == null) {
            eventListeners = new LinkedList<>();
        }
        else {
            eventListeners.clear();
        }

        View view = inflater.inflate(R.layout.fragment_message, container, false);
        // начальная инициализация списка
        // получаем элемент ListView
        messageList = (ListView) view.findViewById(R.id.messageList);
        // создаем адаптер
        context = view.getContext();

        mAuth = FirebaseAuth.getInstance();

        DatabaseReference myRef = FirebaseDatabase.getInstance().getReference("users").child(mAuth.getUid());

        Button button = (Button) view.findViewById(R.id.send_message);
        button.setOnClickListener(this);

        messageText=(TextView)view.findViewById(R.id.message_text) ;
        messageText.setText(MainActivity.savedMessage);
        myRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if(dataSnapshot.getValue() != null) {
                    userStreetId = dataSnapshot.child("street_id").getValue().toString();
                    userBuildingId = dataSnapshot.child("building_id").getValue().toString();

                    notifyEventListeners(new MyEvent(this, MyEvent.Type.UpdateAddressID));
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

        this.addEventListener(new MyEventListener() {
            @Override
            public void processEvent(MyEvent event) {
                if(event.getSource() == null || event.getType() == null) {
                    return;
                }

                switch (event.getType()) {
                    case UpdateAddressID: {

                        if(myRef2 != null) {
                            myRef2.removeEventListener(listenerBuilding);
                        }

                        myRef2 = FirebaseDatabase.getInstance().getReference("streets").child(userStreetId)
                                .child("buildings").child(userBuildingId).child("organization_id");

                        listenerBuilding = new ValueEventListener() {
                            @Override
                            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                organizationId = dataSnapshot.getValue().toString();

                                notifyEventListeners(new MyEvent(this, MyEvent.Type.UpdateOrganizationId));
                            }

                            @Override
                            public void onCancelled(@NonNull DatabaseError databaseError) {

                            }
                        };

                        myRef2.addValueEventListener(listenerBuilding);
                    } break;

                    case UpdateOrganizationId: {

                        if(myRef3 != null) {
                            myRef3.removeEventListener(listenerMessages);
                        }

                        myRef3 = FirebaseDatabase.getInstance().getReference("organization")
                                .child(organizationId).child("messages").child(mAuth.getUid());

                        listenerMessages = new ValueEventListener() {
                            @Override
                            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                messageSnapshot = dataSnapshot;

                                notifyEventListeners(new MyEvent(this, MyEvent.Type.UpdateNews));
                            }

                            @Override
                            public void onCancelled(@NonNull DatabaseError databaseError) {

                            }
                        };

                        myRef3.addValueEventListener(listenerMessages);
                    } break;

                    case UpdateNews: {
                        updateUI();
                        System.gc();
                    } break;
                }
            }
        });

        return view;
    }



    @Override
    public void onClick(View v) {
        messageText.clearFocus();
        //скрыть клавиатуру
        InputMethodManager imm = (InputMethodManager)context.getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(messageText.getWindowToken(), 0);

        switch (v.getId()) {
            case R.id.send_message:
                sendMessage();
                break;
        }
    }

    private void sendMessage()
    {
        if (messageText.getText().toString()==null)
            return;
        if (messageText.getText().toString().trim().length() == 0 )
            return;
        MessageBD mes=new MessageBD(0, messageText.getText().toString(), (new Date()).toString());
        int messageId=messages.size()+1;
        messageText.setText("");
        MainActivity.savedMessage="";
        FirebaseDatabase.getInstance().getReference("organization")
             .child(organizationId).child("messages")
                .child(mAuth.getUid()).child(Integer.toString(messageId)).setValue(mes);
    }

    private void updateUI(){
        messages.clear();
        String userName;

        if(messageSnapshot != null) {
            for (DataSnapshot n : messageSnapshot.getChildren()) {
                if (Integer.parseInt(n.child("income").getValue().toString())==0)
                    userName="Вы:";
                else {
                    userName = "Управляющая компания:";
                    FirebaseDatabase.getInstance().getReference("organization")
                            .child(organizationId).child("messages")
                            .child(mAuth.getUid()).child(n.getKey()).child("read").setValue(1);
                }
                messages.add(new Message(userName, n.child("Text").getValue().toString(), n.child("date").getValue().toString()));
            }
        }
        Collections.reverse(messages);
        MessageAdapter messageAdapter = new MessageAdapter(getActivity(), R.layout.list_message, messages);
        // устанавливаем адаптер
        messageList.setAdapter(messageAdapter);
    }

    @Override
    public void onDestroyView()
    {
        super.onDestroyView();
        eventListeners.clear();
        if(myRef2 != null) {
            myRef2.removeEventListener(listenerBuilding);
        }
        if(myRef3 != null) {
            myRef3.removeEventListener(listenerMessages);
        }
        listenerBuilding=null;
        listenerMessages=null;
    }
}

File: app/src/main/java/com/dvor/my/mydvor/message/MessageFragment.kt
package com.dvor.my.mydvor.message

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.ListView
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.dvor.my.mydvor.*
import com.dvor.my.mydvor.R
import com.dvor.my.mydvor.data.Message
import com.dvor.my.mydvor.data.MessageBD
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import java.util.*

class MessageFragment : Fragment(), View.OnClickListener {

    private lateinit var mAuth: FirebaseAuth
    internal var userStreetId: String = ""
    internal var organizationId: String = ""
    internal var userBuildingId: String = ""
    internal var myRef2: DatabaseReference? = null
    internal var myRef3: DatabaseReference? = null
    internal var messageSnapshot: DataSnapshot? = null
    internal var listenerBuilding: ValueEventListener? = null
    internal var listenerMessages: ValueEventListener? = null

    private val messages = ArrayList<Message>()
    private var messageText: TextView? = null
    private lateinit var messageList: ListView
    internal lateinit var context: Context

    private fun addEventListener(eventListener: MyEventListener) {
        eventListeners!!.add(eventListener)
    }

    fun notifyEventListeners(event: MyEvent) {
        for (eventListener in eventListeners!!) {
            eventListener.processEvent(event)
        }
    }

    //сохранение текста неотправленного сообщения
    //костыльно но работает
    override fun onSaveInstanceState(outState: Bundle) {
        if (messageText == null) {
            outState.putString("message", MainActivity.savedMessage)
        } else {
            outState.putString("message", messageText!!.text.toString())
            MainActivity.savedMessage = messageText!!.text.toString()
        }
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        super.onCreate(savedInstanceState)

        if (eventListeners == null) {
            eventListeners = LinkedList()
        } else {
            eventListeners!!.clear()
        }

        val view = inflater.inflate(R.layout.fragment_message, container, false)
        // начальная инициализация списка
        // получаем элемент ListView
        messageList = view.findViewById(R.id.messageList)
        // создаем адаптер
        context = view.context

        mAuth = FirebaseAuth.getInstance()

        val myRef = FirebaseDatabase.getInstance().getReference("users").child(mAuth.uid!!)

        val button = view.findViewById<Button>(R.id.send_message)
        button.setOnClickListener(this)

        messageText = view.findViewById(R.id.message_text)
        messageText!!.text = MainActivity.savedMessage
        myRef.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.value != null) {
                    userStreetId = dataSnapshot.child("street_id").value!!.toString()
                    userBuildingId = dataSnapshot.child("building_id").value!!.toString()

                    notifyEventListeners(MyEvent(this, Type.UpdateAddressID))
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })

        this.addEventListener(object : MyEventListener {
            override fun processEvent(event: MyEvent) {
                if (event.source == null) {
                    return
                }

                when (event.type) {
                    Type.UpdateAddressID -> {

                        if (myRef2 != null) {
                            myRef2!!.removeEventListener(listenerBuilding!!)
                        }

                        myRef2 = FirebaseDatabase.getInstance().getReference("streets").child(userStreetId)
                                .child("buildings").child(userBuildingId).child("organization_id")

                        listenerBuilding = object : ValueEventListener {
                            override fun onDataChange(dataSnapshot: DataSnapshot) {
                                organizationId = dataSnapshot.value!!.toString()

                                notifyEventListeners(MyEvent(this, Type.UpdateOrganizationId))
                            }

                            override fun onCancelled(databaseError: DatabaseError) {

                            }
                        }

                        myRef2!!.addValueEventListener(listenerBuilding!!)
                    }

                    Type.UpdateOrganizationId -> {

                        if (myRef3 != null) {
                            myRef3!!.removeEventListener(listenerMessages!!)
                        }

                        myRef3 = FirebaseDatabase.getInstance().getReference("organization")
                                .child(organizationId).child("messages").child(mAuth.uid!!)

                        listenerMessages = object : ValueEventListener {
                            override fun onDataChange(dataSnapshot: DataSnapshot) {
                                messageSnapshot = dataSnapshot

                                notifyEventListeners(MyEvent(this, Type.UpdateNews))
                            }

                            override fun onCancelled(databaseError: DatabaseError) {

                            }
                        }

                        myRef3!!.addValueEventListener(listenerMessages!!)
                    }

                    Type.UpdateNews -> {
                        updateUI()
                        System.gc()
                    }
                }
            }
        })

        return view
    }


    override fun onClick(v: View) {
        messageText!!.clearFocus()
        //скрыть клавиатуру
        val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(messageText!!.windowToken, 0)

        if (v.id == R.id.send_message) {
            sendMessage()
        }
    }

    private fun sendMessage() {
        if (messageText!!.text.toString().trim { it <= ' ' }.isEmpty())
            return
        val mes = MessageBD(0, messageText!!.text.toString(), Date().toString())
        val messageId = messages.size + 1
        messageText!!.text = ""
        MainActivity.savedMessage = ""
        FirebaseDatabase.getInstance().getReference("organization")
                .child(organizationId).child("messages")
                .child(mAuth.uid!!).child(messageId.toString()).setValue(mes)
    }

    private fun updateUI() {
        messages.clear()
        var userName: String

        if (messageSnapshot != null) {
            for (n in messageSnapshot!!.children) {
                if (Integer.parseInt(n.child("income").value!!.toString()) == 0)
                    userName = "Вы:"
                else {
                    userName = "Управляющая компания:"
                    FirebaseDatabase.getInstance().getReference("organization")
                            .child(organizationId).child("messages")
                            .child(mAuth.uid!!).child(n.key!!).child("read").setValue(1)
                }
                messages.add(Message(userName, n.child("text").value.toString(), n.child("date").value.toString()))
            }
        }
        messages.reverse()
        val messageAdapter = MessageAdapter(activity, R.layout.list_message, messages)
        // устанавливаем адаптер
        messageList.adapter = messageAdapter
    }

    override fun onDestroyView() {
        super.onDestroyView()
        eventListeners!!.clear()
        if (myRef2 != null) {
            myRef2!!.removeEventListener(listenerBuilding!!)
        }
        if (myRef3 != null) {
            myRef3!!.removeEventListener(listenerMessages!!)
        }
        listenerBuilding = null
        listenerMessages = null
    }

    companion object {

        private var eventListeners: MutableList<MyEventListener>? = null
    }
}